specs/CustomDownload.md

```diff
@@ -0,0 +1,528 @@
+    out with an explanation either here or in the Examples
+-->
+
+
+# API Notes
+<!-- TEMPLATE
+    Option 1: Give a one or two line description of each API (type and member),
+        or at least the ones that aren't obvious from their name. These
+        descriptions are what show up in IntelliSense. For properties, specify
+        the default value of the property if it isn't the type's default (for
+        example an int-typed property that doesn't default to zero.)
+
+    Option 2: Put these descriptions in the below API Details section,
+        with a "///" comment above the member or type.
+-->
+See [API Details](#api-details) section below for API reference.
+
+# API Details
+<!-- TEMPLATE
+    The exact API, in IDL format for our COM API and
+    in MIDL3 format (https://docs.microsoft.com/en-us/uwp/midl-3/)
+    when possible, or in C# if starting with an API sketch for our .NET and WinRT API.
+
+    Include every new or modified type but use // ... to remove any methods,
+    properties, or events that are unchanged.
+
+    (GitHub's markdown syntax formatter does not (yet) know about MIDL3, so
+    use ```c# instead even when writing MIDL3.)
+
+    Example:
+
+    ```
+    /// Event args for the NewWindowRequested event. The event is fired when content
+    /// inside webview requested to open a new window (through window.open() and so on.)
+    [uuid(34acb11c-fc37-4418-9132-f9c21d1eafb9), object, pointer_default(unique)]
+    interface ICoreWebView2NewWindowRequestedEventArgs : IUnknown
+    {
+        // ...
+
+        /// Window features specified by the window.open call.
+        /// These features can be considered for positioning and sizing of
+        /// new webview windows.
+        [propget] HRESULT WindowFeatures([out, retval] ICoreWebView2WindowFeatures** windowFeatures);
+    }
+    ```
+
+    ```c# (but really MIDL3)
+    public class CoreWebView2NewWindowRequestedEventArgs
+    {
+        // ...
+
+	       public CoreWebView2WindowFeatures WindowFeatures { get; }
+    }
+    ```
+-->
+## Win32 C++
+```c#
+// Enums and structs
+
+typedef enum COREWEBVIEW2_DOWNLOAD_DANGER_TYPE {
+  // The download is safe.
+  DANGER_TYPE_NONE,
+  // The download file is dangerous.
+  DANGER_TYPE_FILE,
+  // The URI leads to a malicious file download.
+  DANGER_TYPE_URI,
+  // The content of the file is dangerous.
+  DANGER_TYPE_CONTENT,
+  // Not enough data to know whether download is dangerous.
+  DANGER_TYPE_UNCOMMON,
+  // The host is known to serve dangerous content.
+  DANGER_TYPE_HOST,
+  // The download is an application or extension that modifies browser or
+  // computer settings.
+  DANGER_TYPE_POTENTIALLY_UNWANTED,
+} COREWEBVIEW2_DOWNLOAD_DANGER_TYPE;
+
+typedef enum COREWEBVIEW2_DOWNLOAD_STATE {
+  /// The download is in progress.
+  DOWNLOAD_IN_PROGRESS,
+  /// The download has been interrupted.
+  DOWNLOAD_INTERRUPTED,
+  /// The download is complete.
+  DOWNLOAD_COMPLETE,
```

Comment:
Is there a distinction between paused and failed?

Reply:
Yes, an interrupted download is a failed download, although in some cases an interrupted download can be resumed. A paused download (paused by user) is in the `DOWNLOAD_IN_PROGRESS` state, not `DOWNLOAD_INTERRUPTED`, and it can always be resumed. Hopefully this is clear in the comments now.

Reply:
If the download is paused by the user and the state doesn't change, how will the app know they should call Resume to resume the download?

Reply:
The app can handle the user pausing and resuming a download through the Pause and Resume methods on the download. The properties `CanResume` and `IsPaused` are also available. Is that good enough?

Reply:
Ideally there's one set of code to deal with the download stopping/interruptions/pausing in the state change event handler rather than also having additional code to handle the user pausing the download and having to separately track that info and having separate code for that.